terra/contracts/nft-bridge/src/contract.rs

```diff
@@ -0,0 +1,602 @@
+    asset: HumanAddr,
+    token_id: String,
+    recipient_chain: u16,
+    recipient: [u8; 32],
+    nonce: u32,
+) -> StdResult<Response> {
+    if recipient_chain == CHAIN_ID {
+        return ContractError::SameSourceAndTarget.std_err();
+    }
+
+    let asset_chain: u16;
+    let asset_address: [u8; 32];
+
+    let cfg: ConfigInfo = config_read(deps.storage).load()?;
+    let asset_canonical: CanonicalAddr = deps.api.addr_canonicalize(&asset)?;
+
+    let mut messages: Vec<CosmosMsg> = vec![];
+
+    if let Ok(_) = wrapped_asset_address_read(deps.storage).load(asset_canonical.as_slice()) {
+        // This is a deployed wrapped asset, burn it
+        messages.push(CosmosMsg::Wasm(WasmMsg::Execute {
+            contract_addr: asset.clone(),
+            msg: to_binary(&cw721_wrapped::msg::ExecuteMsg::Burn::<Option<Empty>> {
+                token_id: token_id.clone(),
+            })?,
+            funds: vec![],
+        }));
+
+        let wrapped_token_info: cw721_wrapped::msg::WrappedAssetInfoResponse = deps
+            .querier
+            .custom_query(&QueryRequest::<Empty>::Wasm(WasmQuery::Smart {
+                contract_addr: asset.clone(),
+                msg: to_binary(&cw721_wrapped::msg::QueryMsg::WrappedAssetInfo {})?,
+            }))?;
+
+        asset_address = wrapped_token_info.asset_address.to_array()?;
+        asset_chain = wrapped_token_info.asset_chain;
+    } else {
+        // Native NFT, lock it up
+        messages.push(CosmosMsg::Wasm(WasmMsg::Execute {
+            contract_addr: asset.clone(),
+            msg: to_binary(&cw721_base::msg::ExecuteMsg::<Option<Empty>>::TransferNft {
+                recipient: env.contract.address.to_string(),
+                token_id: token_id.clone(),
+            })?,
+            funds: vec![],
+        }));
+
+        asset_chain = CHAIN_ID;
+        asset_address = extend_address_to_32_array(&asset_canonical);
+    };
+
+    let cw721::ContractInfoResponse { symbol, name, .. } =
+        deps.querier
+            .custom_query(&QueryRequest::<Empty>::Wasm(WasmQuery::Smart {
+                contract_addr: asset.clone(),
+                msg: to_binary(&cw721_base::msg::QueryMsg::ContractInfo {})?,
+            }))?;
+
+    let cw721::NftInfoResponse::<Option<Empty>> { token_uri, .. } =
+        deps.querier
+            .custom_query(&QueryRequest::<Empty>::Wasm(WasmQuery::Smart {
+                contract_addr: asset.clone(),
+                msg: to_binary(&cw721_base::msg::QueryMsg::NftInfo {
+                    token_id: token_id.clone(),
+                })?,
+            }))?;
+
+    let transfer_info = TransferInfo {
+        nft_address: asset_address,
+        nft_chain: asset_chain,
+        symbol: string_to_array(&symbol),
+        name: string_to_array(&name),
+        token_id: to_external_token_id(deps.storage, asset_chain, &asset_address, token_id)?,
+        uri: BoundedVec::new(token_uri.unwrap_or("".to_string()).into())?,
```

Comment:
why would this fail? shouldn't we abort the transfer if tokenURI is not cleanly returned?

Reply:
My thinking was that since token_uri is an optional extension to CW721, we might legitimately have NFTs that don't have that field, and still want to support them. Should we reject them instead?

Reply:
Let's loop in @evan-gray. Probably confusing if NFTs show up as blank on the other side.

Reply:
Fair, although that should only happen to NFTs that were blank to begin with. That being said, since I wrote this code, I've seen that the other bridge contracts assume the existence of the token_uri so there isn't much point in deviating from them. I'll change it so it produces an appropriate error message